lib/config/formatter.go

```diff
@@ -163,14 +165,75 @@ func (tf *textFormatter) Format(e *log.Entry) ([]byte, error) {
 	// write caller last if enabled
 	if tf.callerEnabled && caller != "" {
 		w.writeField(caller, noColor)
-
 	}
 
 	w.WriteByte('\n')
 	data = w.Bytes()
 	return data, nil
 }
 
+// jsonFormatter implements the logrus.Formatter interface and adds extra
+// fields to log entries
+type jsonFormatter struct {
+	log.JSONFormatter
+
+	callerEnabled    bool
+	componentEnabled bool
+}
+
+// CheckAndSetDefaults checks and sets log format configuration
+func (j *jsonFormatter) CheckAndSetDefaults(extraFields []string) error {
+	// set log formatting
+	if extraFields == nil {
+		extraFields = KnownFormatFields.names()
+	}
+
+	// parse input
+	res, err := parseInputFormat(extraFields)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if utils.SliceContainsStr(res, timestampField) {
+		j.JSONFormatter.DisableTimestamp = true
+	}
+
+	if utils.SliceContainsStr(res, callerField) {
+		j.callerEnabled = true
+	}
+
+	if utils.SliceContainsStr(res, componentField) {
+		j.componentEnabled = true
+	}
+
+	// rename default fields
+	j.JSONFormatter = log.JSONFormatter{
+		FieldMap: log.FieldMap{
+			log.FieldKeyTime:  timestampField,
+			log.FieldKeyLevel: levelField,
+			log.FieldKeyMsg:   messageField,
+		},
+	}
+
+	return nil
+}
+
+// Format implements logrus.Formatter interface
+func (j *jsonFormatter) Format(e *log.Entry) ([]byte, error) {
+	if j.callerEnabled {
+		path := formatCallerWithPathAndLine()
+		e.Data[callerField] = path
+	}
+
+	if j.componentEnabled {
```

Comment:
When we construct the log entry for `text`, we always include the message field so should we keep the `json` format consistent with that or change the `text` format to only print the message if specified? Or does it matter? 

https://github.com/gravitational/teleport/blob/master/lib/config/formatter.go#L155-L157

Reply:
I am sorry I don't see how's this related to the message field. As far as I remember the message field will always be added to the logs here too

Reply:
You're right, I see that now